(QENG-3370) Add pe_* value CLI override options.

waynr opened the pull request.

waynr commented:
@doug-rosser how does this look?

Also, please note that this patch still needs tests added.

rick commented on test/test_generator.rb:
```diff
@@ -0,0 +1,47 @@
+      end
+
+      fixture_file = File.open(f, "r")
+      fixture_hash = YAML.load(fixture_file)
```
Idiomatically, I would probably drop the temp file variable and go with:

``` ruby
fixture_hash = YAML.load_file(f)
```

or

``` ruby
fixture_hash = YAML.load(File.read(f))
```

(The "or" because I can't remember if `load_file` is always available, may depend on YAML version or implementation).

rick commented:
OK, so the fixtures are used to snapshot the behavior of the generator at this point in time?

rick commented on test/test_generator.rb:
```diff
@@ -0,0 +1,47 @@
+      options = fixture_hash["options_string"]
+      options = options.split
+      test_hash = run_cli_with_options(options)
+      assert_equal(fixture_hash["expected_hash"], test_hash)
```
This is probably a good place to document (for instance via the 3rd argument to `assert_equal`) what this failure means. I believe it means "the behavior of the generator has changed from the snapshotted behavior".

If that's true folks are also going to want guidance on how/when to re-snapshot things so that they can take the changed behavior as a new baseline.

waynr commented:
@rick yeah, snapshotting is a good way to describe it. I think I've addressed all your comments. I added a `generate:fixtures` rake task and added an `assert_equal` mismatch helper message indicating which fixture borked and the options passed to `beaker-hostgenerator`.

Unfortunately the tests stop running on the first assertion failure so I'm still not sure this is a particularly good way to write the `test_fixtures` function. It'd be nice to see all the broken fixtures for a given `beaker-hostgenerator` change at once.

doug-rosser commented:
@waynr The code looks good, but the old (weird) behavior is not present. Even though I think the usefulness is dubious, the old code (master branch) would set pe_dir if pe_family and pe_version were the same string. This new code doesn't preserve that behavior. Example:
`[beaker-hostgenerator (master)]$ env | grep pe_
pe_family=2016.1.0
pe_version=2016.1.0
[beaker-hostgenerator (master)]$ bin/beaker-hostgenerator centos6-64mdca
WARNING: Starting with beaker-hostgenerator 1.x platform strings for "el" hosts
will correspond to the actual linux distribution name. ie, the platform string
corresponding to a host specified as "centos4_64a" will be "centos-4-x86_64"
rather than "el-4-x86_64". It is recommended that you update your project's test
suites ASAP or be forced to do so when beaker-hostgenerator development moves on
to the 1.x series. We don't intend to backport features or platforms to 0.x.
---
HOSTS:
  centos6-64-1:
    pe_dir: http://enterprise.delivery.puppetlabs.net/archives/releases/2016.1.0/
    pe_ver: 2016.1.0
    pe_upgrade_dir: 
    pe_upgrade_ver: 
    hypervisor: vmpooler
    platform: el-6-x86_64
    template: centos-6-x86_64
    roles:
    - agent
    - master
    - database
    - dashboard
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
[beaker-hostgenerator (master)]$ 
[beaker-hostgenerator (master)]$ git checkout qeng-3370 
Switched to branch 'qeng-3370'
Your branch is up-to-date with 'upstream/qeng-3370'.
[beaker-hostgenerator (qeng-3370)]$ 
[beaker-hostgenerator (qeng-3370)]$ bin/beaker-hostgenerator centos6-64mdca
WARNING: Starting with beaker-hostgenerator 1.x platform strings for "el" hosts
will correspond to the actual linux distribution name. ie, the platform string
corresponding to a host specified as "centos4_64a" will be "centos-4-x86_64"
rather than "el-4-x86_64". It is recommended that you update your project's test
suites ASAP or be forced to do so when beaker-hostgenerator development moves on
to the 1.x series. We don't intend to backport features or platforms to 0.x.
---
HOSTS:
  centos6-64-1:
    pe_dir: 
    pe_ver: 
    pe_upgrade_dir: 
    pe_upgrade_ver: 
    hypervisor: vmpooler
    platform: el-6-x86_64
    template: centos-6-x86_64
    roles:
    - agent
    - master
    - database
    - dashboard
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
`

waynr commented:
@doug-rosser okay, I've fixed that bug. The problem was here: https://github.com/puppetlabs/beaker-hostgenerator/commit/aa3e08cdca7bc40e9eae037ac625e7c5fc6acc22#diff-0623c137fe4358c047889a4ff1adc01bL62

There were some other module constant initialization problems throughout the beaker-hostgenerator library that were leading to incorrectly-generated fixtures which have now been fixed.

doug-rosser commented:
:+1: :+1: :+1:

nwolfe commented on .projectile:
```diff
@@ -0,0 +1 @@
+-/vendor
```
Do we want to commit an emacs file like this?

waynr replied:
I added it because I think of it as similar to a `.gitignore` file which can often contain subjectively-named file paths. e.g. https://github.com/puppetlabs/beaker-hostgenerator/blob/master/.gitignore#L13-L14 (That `.gitignore` originally came from puppetlabs/beaker-template)

The difference here is that I don't care as much about `.gitignore` so I never bothered to add `vendor` to it the way I do for projectile. Local package installation is much more annoying in projectile than in git.

waynr replied:
I would defer to you as far as whether you think it's a good idea or not--I haven't really thought through the implications for other emacs users. I'm sure there is some way to configure projectile exclude paths in one's emacs configuration which would addres my problem more generally.

nwolfe replied:
I think you want to add `.projectile` to your global `~/.gitignore`, and then you can sprinkle these projectile ignore files anywhere.

nwolfe replied:
Alternately, you could just configure projectile to globally ignore `vendor` directories with `projectile-globally-ignored-files` if that suits you better.  

Either way I wouldn't think we'd want to start committing editor-specific files.

waynr replied:
Are you saying that I should remove this file from the PR? I don't have a problem with that, but just to clarify I intentially added it in the first place.

nwolfe replied:
Ha yeah I'd say remove it from the PR and achieve what you're trying to do using one of the options I mentioned above.

But this is a good idea, I'm going to ignore vendor directories too :smile:

rick commented on README.md:
```diff
@@ -86,6 +101,82 @@ CONFIG:
+`generate:fixtures` task:
+
+- When you modify the `FixtureGenerator` to generate new fixtures.
+- When you need to fix bug (generated hosts are not usable without your change,
```
:lipstick: "fix a bug"

waynr replied:
Fixed!

nwolfe commented on Gemfile:
```diff
@@ -21,6 +21,8 @@ group :acceptance_testing do
   gem "beaker", *location_for(ENV['BEAKER_VERSION'] || '~> 2.0')
 end
 
+gem "minitest"
```
Should this be under a test or development group so it's not required for normal usage?

nwolfe commented on README.md:
```diff
@@ -86,6 +101,82 @@ CONFIG:
+figure out what you did wrong and figure out how to achieve your goal without
+potentially breaking `beaker-hostgenerator` for other users.
+
+There are are few circumstances when you should expect to run the
```
s/are are/are

nwolfe commented on test/util/generator_helpers.rb:
```diff
@@ -0,0 +1,150 @@
+  include BeakerHostGenerator::Data
+  include BeakerHostGenerator::Data::Vmpooler
+
+  def run_cli_with_options options = []
```
We should put the parens around the method parameters here and `open_file` below, for readability and consistency

nwolfe commented on test/util/generator_helpers.rb:
```diff
@@ -0,0 +1,150 @@
+    yaml_string = cli.execute
+
+    hash = YAML.load(yaml_string)
+    return hash
```
Whoa, forgot this was a keyword! Looks like most code doesn't bother using `return`, I'd probably remove it but no big deal

waynr replied:
I would prefer to leave it in and deal with it in a more comprehensive rubocop-driven style refactor. I guess the implied return value of the last expression in the function is probably pretty familiar for you coming from clojure and all...I happen to prefer explicit over implicit for behaviors like this (coming from C/bash/python world) but will definitely be changing this in a later cleanup PR.

nwolfe commented on test/util/generator_helpers.rb:
```diff
@@ -0,0 +1,150 @@
+    [
+      {'case_name' => 'pe_version_and_pe_family',
+       'environment_variables' => {
+         'pe_version' => '6.6.6',
```
:metal:

nwolfe commented:
:+1: - left some minor nitpicks

er0ck commented on Rakefile:
```diff
@@ -37,10 +48,25 @@ A quick acceptance test, named because it has no pre-suites to run
 # these are the default tasks invoked when only the namespace is referenced.
 # they're needed because `task :default` in those blocks doesn't work as expected.
-task 'test:spec' => 'test:spec:run'
+task 'test:spec' => ['test:spec:run', 'test:spec:minitest']
```
shouldn't :generate:fixtures be a dependency of this task?

waynr replied:
No, see the documentation at:
https://github.com/puppetlabs/beaker-hostgenerator/blob/qeng-3370/README.md#generated-fixtures

The fixture generator is not meant to be run every time the tests are run--that's why the fixtures are committed to the repo, to "snapshot" the current behavior of beaker-hostgenerator in a way that helps prevent regressions as we refactor or repurpose parts of the code.

Anyway, we generated fixtures before every time we run the specs that use those fixtures that spec would always pass unless we do something complicated to generate the fixtures for some older commit of the repo then test using them against the HEAD of the PR branch (which would assume a certain CI setup).

er0ck commented on lib/beaker-hostgenerator/data.rb:
```diff
@@ -1,11 +1,21 @@
+
+    def pe_upgrade_family
+      ENV['pe_upgrade_family']
+    end
```
any reason these are lowercase?
also, beaker prepends these with BEAKER_
should we do something similar here?

waynr replied:
Historicaly reasons mostly:
https://github.com/puppetlabs/sqa-utils-gem/blob/master/bin/genconfig#L18-L22

I'm definitely willing to consider changing their names for beaker-hostgenerator 1.x but doing so will have implications throughout CI and probably user scripts.

er0ck commented on lib/beaker-hostgenerator/data.rb:
```diff
@@ -51,5 +61,7 @@ module Data
       }
     }
+
+    module_function :pe_version, :pe_family, :pe_upgrade_version, :pe_upgrade_family
```
the only place i see this module used is: test/util/generator_helpers.rb
why not just mix this module in?  particularly if the above file ends up with multiple modules/classes.

waynr replied:
I don't know, nothing in Ruby makes sense to me and whenever I think too hard about it I become immaturely angry with myself for working a job that requires me to write a language that I consider to be bad programming practice. (Things I tend to think of as good programming practice often turn out to be "ineloquent" in the duby community.)

On a more reasonable note, I may have done this back when I tried calling these functions when setting values for a module constant here:
https://github.com/puppetlabs/beaker-hostgenerator/blob/master/lib/beaker-hostgenerator/generator.rb#L14

As it turns out, the use of a module constant whose value ultimately is set programatically and can change in different circumstances is actually quite a bad idea (hopefully I've learned my lesson). If I dwell too much on this point I am fairly certain I will become irrationally angry that ruby allows module constants to be set with functions or even to values that can be mutated at all.

TL;DR I think the use of mixins is reasonable so I'll go ahead and do that, thanks for the feedback!

waynr replied:
This has been fixed.

er0ck commented on lib/beaker-hostgenerator/util.rb:
```diff
@@ -22,7 +22,7 @@ def pe_dir(version, family)
     def fixup_node(cfg)
       # PE 2.8 doesn't exist for EL 4. We use 2.0 instead.
-      if cfg['platform'] =~ /el-4/ and PE_VERSION =~ /2\.8/
+      if cfg['platform'] =~ /el-4/ and BeakerHostGenerator::Data.pe_version =~ /2\.8/
```
should this kind of biz data go in another, single place?

waynr replied:
Yes, but changing it will require a major version bump. (It is supported here for backwards compatibility with https://github.com/puppetlabs/sqa-utils-gem/blob/master/bin/genconfig#L259-L264)

waynr replied:
Unless you were suggesting that it go elsewhere in the beaker-hostgenerator library (my above comment assumes you meant that it should be somewhere outside this repo).

er0ck commented on test/fixtures/default/arista4-32a:
```diff
@@ -0,0 +1,20 @@
+    nfs_server: none
+    consoleport: 443
+    pooling_api: http://vmpooler.delivery.puppetlabs.net/
+expected_exception: 
```
should all these fixtures be checked in?  are these not what are being generated by the code above?

justinstoller commented on lib/beaker-hostgenerator/data.rb:
```diff
@@ -1,11 +1,21 @@
-    PE_UPGRADE_VERSION=ENV['pe_upgrade_version']
-    PE_UPGRADE_FAMILY=ENV['pe_upgrade_family']
+    def pe_version
+      ENV['pe_version']
```
Do we want to memoize calls to the environment here or do we expect these values to change?

waynr replied:
We expect these values to change in the context of both test fixture generation and test fixture validation:
https://github.com/puppetlabs/beaker-hostgenerator/blob/qeng-3370/spec/beaker-hostgenerator/generator_spec.rb#L106-L113
https://github.com/puppetlabs/beaker-hostgenerator/blob/qeng-3370/test/util/generator_helpers.rb#L34-L40

Otherwise I probably would probably hate Ruby slightly less right now.

justinstoller commented:
In general this looks like a good change and it has great documentation and automation around it.

However, I usually consider fixture/data driven testing like this an anti-pattern. I think it has a valuable place when you have legacy behavior around spaghetti code that is hard to test otherwise, or you have folks writing tests that may not be familiar with the code under test and/or development in general. I would prefer testing the _logic_ in the code not the variations in the data.

Having said that, that's the only thing I have an issue with, and it would appear from the previous comments that I'm in the minority, just felt obligated to mention it.

All in all, :+1:

er0ck commented:
justin has a good point about what appears to be feature level testing.  i'm okay with that. maybe should move to an integration suite, and unit tests should be added for unit/class/module logic
